cdrwtool: Fix FORMAT UNIT

scdbackup opened the pull request:
Fixes: https://github.com/pali/udftools/issues/33

pali commented:
Automated tests compiled your patches with more gcc, clang and tcc versions. And compilatin failed. You can look at compile logs from each test via blue hyperlink "Details" in "All checks have failed" box. Please fix patches so they can be compiled.

scdbackup commented:
Hi,

i locally removed the two unused variables and hacked my local
./cdrwtool/Makefile so that it uses -Wall -Werror for compiling like
Travis CI does. No further problems found.

What to do now ? Commit and push ? How to update the pull request ?

Have a nice day :)

Thomas

pali commented:
Pull request is reflecting status of your branch master from your repository https://github.com/scdbackup/udftools. So just update commits in your repository and push changes to your repository via `git push --force <arguments>`.

pali commented on cdrwtool/cdrwtool.c:
```diff
@@ -416,11 +422,55 @@ int blank_disc(int fd, int type)
 
-	print_completion_info(fd);
-	return 0;
+	return wait_for_unit_ready(fd, "BLANK", WAIT_BLANK_SECONDS);
```
I would suggest to pass timeout in hz units like in `wait_cmd` and `wait_cmd_sense` functions. Just to use parameters in these functions in the same unit. So it would not be suspicous that once variable `timeout` is in seconds and once in hz.

scdbackup commented:
Hi,

> push changes to your repository via git push --force <arguments>.

I used the same command "hub push scdbackup" as before and the commit 
shows up here.

It won't work for a long time, though. I got a mail from GitHub with the
warning that password based pushing will be banned soon.

(I get ecological scruples seeing Travis running dozens of compile jobs 
after a tiny and unspecific commit. All the little Joules wasted. My 
local CPU is at 3800 MHz just from watching.)

Have a nice day :)

Thomas

pali commented on cdrwtool/cdrwtool.c:
```diff
@@ -73,80 +72,87 @@ void dump_sense(unsigned char *cdb, struct request_sense *sense)
+	memset(&cgc, 0, sizeof(cgc));
+	memset(buffer, 0, sizeof(buffer));
+
+	while (total_elapsed < timeout) {
```
... and then just this place needs to be update to convert `timeout` from hz to `timeout_sec` in seconds.

pali commented:
> It won't work for a long time, though. I got a mail from GitHub with the warning that password based pushing will be banned soon.

You can extending token access on https://github.com/settings/tokens so that it would cover also push permission or you can upload your public ssh key to github at https://github.com/settings/keys or directly at https://github.com/settings/ssh/new and use ssh protocol for pushes.

scdbackup commented:
Hi,

the "Reply" fields underneath your code reviews do not accept input from
my browser. So as comment:

The loop in wait_for_unit_ready() counts tries, not really seconds.
The relation to time is merely by the call of sleep(1).

I could make it really time dependent, but actually the timeout value
of 3600 seconds is so deliberate that it does not matter much, whether
the execution times of TEST UNIT READY are counted or not.
After all, you won't get your drive back before it is ready again,
regardless whether cdrwtool stops waiting or not.

I googled "how to get linux hz in userspace" and got the impression that 
sysconf(_SC_CLK_TCK) might yield the HZ value in userspace. But it is 
not documented that way. The man page talks of "clock ticks".
So i'd like to stay with seconds ~= tries.
wait_cmd_sense() (formerly wait_cmd()) hands over a kernel-ish time value
to ioctl(CDROM_SEND_PACKET). Thus the HZ, i guess.
It does not interpret the value and thus cannot do it wrong as long as
bitrot does not strike.

Have a nice day :)

Thomas

pali commented:
Yes, it is `HZ` macro which you have already used/touched in cdrwtool.h file when defining `WAIT_BLANK_SECONDS` and `WAIT_BLANK`.

I mean to use just `WAIT_BLANK` macro and divide its value by `HZ` on places where you are currently using units in seconds.

scdbackup commented:
Hi,

> I mean to use just WAIT_BLANK macro and divide its value by HZ on places
> where you are currently using units in seconds.

Duh. (The snot level must have risen again.)

I will re-run the drive tests and then commit.

Have a nice day :)

Thomas

pali commented:
When updating, could you "squash" those fixup changes into relevant first and second commits? `git rebase -i <first_commit>~1` should allow you to "squash" two (or more commits) into one, including changing their order, etc...

pali commented on cdrwtool/cdrwtool.c:
```diff
@@ -73,80 +72,87 @@ void dump_sense(unsigned char *cdb, struct request_sense *sense)
+	}
+	fprintf(stderr,
+		"ERROR: Drive not ready more than %d seconds after %s\n",
+		timeout, what);
```
Here is also value `timeout` in hz units, but message say _seconds_. So `timeout / HZ` should be used.

scdbackup commented:
Hi,

i don't feel able to learn new git tricks today. If you want i will wrap
it all up freshly when we are done with the review. But currently i only
stumble over my head.

-------------------------------------------------------------------------

Whatever, while scanning for more "timeout" i found a default in
cmd_wait_sense() which obviously speculates about the value of HZ:

            if (timeout <= 0)
                    timeout = 500;

This is inherited from the old cmd_wait().

Do you agree that this should be some (xyz * HZ) or one of the WAIT_*
values from cdrwtool.h ?
If so: what xyz or which WAIT_* would you prefer ?

-------------------------------------------------------------------------

Now i remember that i wanted to simulate failure and timeout in the
wait_for_unit_ready() loop in order to test the behavior. But not today.

Have a nice day :)

Thomas

pali commented:
> If you want i will wrap it all up freshly when we are done with the review.

Changes are fine now. Just squash fixups into the first and second commit. And will merge it.

And about other timeouts, let it for other future pull request / improvement.

scdbackup commented:
Hi,

i fear that i should have known about the wish to squash before
committing the polishing fixes:

    $ git rebase -i 8b93876bc79be23674f5f14a143216ea444332b5~1
    
with

    pick 8b93876bc79be23674f5f14a143216ea444332b5 cdrwtool: Revive USE_IMMED
    squash 130b86e1b265c7cdb6b9b6fa4eab57919c3c9473 cdrwtool: Remove two unused variables
    squash 42f8d54a1125a39be065269908b7a2cc32e907e9 Give up newly introduced macro WAIT_BLANK_SECONDS
    squash d145cddc0bb3fc912004120ce75abf2a3ffeeb13 Correct time value that is told by timeout message
    pick 474e01c3d1a226a99a367dc23f5fd4193b3d2f3d cdrwtool: Switch to FORMAT UNIT Format Code 1

yields

    Auto-merging cdrwtool/cdrwtool.c
    CONFLICT (content): Merge conflict in cdrwtool/cdrwtool.c
    error: could not apply 130b86e... cdrwtool: Remove two unused variables
    Resolve all conflicts manually, mark them as resolved with
    "git add/rm <conflicted_files>", then run "git rebase --continue".
    You can instead skip this commit: run "git rebase --skip".
    To abort and get back to the state before "git rebase", run "git rebase --abort".
    Could not apply 130b86e... cdrwtool: Remove two unused variables

I guess it is because the first and the second polisher touch the realm
of both original commits.

How about i adapt my patches (they are static and cannot complain),
apply and commit them to a new branch stemming from your last commit
(aef7f01), make sure the resulting files match those in the master branch,
and open a new pull request ?

Have a nice day :)

Thomas

pali commented:
Rebasing is aborted by `git rebase --abort` and it revert it into state at which it was before rebasing.

Maybe simplier for you it would be to do `git reset 8b93876bc79be23674f5f14a143216ea444332b5~1` which "uncommit" your all changes, and let it them in your working tree. Then via `git add -p` followed by `git commit` you can selectively choose which changes go into first commit and repeating these steps allows you to create also second commit.

scdbackup commented:
Hi,

> Rebasing is aborted by git rebase --abort and it revert it into state at
> which it was before rebasing.

No rewinding was needed. It happened when i practiced in a temporary branch
which i then disposed by "git branch -D". The master branch is undamaged.

> Maybe simplier for you it would be to do git reset
> 8b93876bc79be23674f5f14a143216ea444332b5~1 which "uncommit"  your all
> changes, and let it them in your working tree. Then via git add -p followed
> by git commit you can selectively choose which changes go into first commit
> and repeating these steps allows you to create also second commit.

Duh ?

I understand from the man page that "git add -p" is about picking
from a patch. I already have two separate patches which are updated to
yield the same result as the five commits in branch "master".
It's just that their two commits now sit in a local branch "v2" which
stems from your commit aef7f01, like "master" does.

Can't i just submit "v2" for pulling ?
Must i really destroy the evidence in "master" ?
Can i rename "master" to "old_master" and "v2" to "master" without
capsizing the pull request ?

Have a nice day :)

Thomas

pali commented:
`git add -p` ask you interactively about each "chunk" if you want to use it for commit. So you say "y"(es) for all changes which you want to put into first commit. Then you call `git commit` which will create a commit from changes on which you said "y"(es). After that you call `git add -p` again and agin create a second commit.

pali commented:
You can rename branch if you want. That is up to you. This pull request will aways track what is stored in branch "master". Therefore if you rename current master to master_old, then create a new master and push changes from new master to "master" on github, then pull request will again handle what is in (new) master. Github does not track any renames, it just see current names.

scdbackup committed:
cdrwtool: Revive USE_IMMED

Some drives get stuck when a FORMAT UNIT without Immed bit command lasts
about 500 seconds. Other drives behave as if the Immed bit was set even if
it is not set.
The current code for waiting after starting an asynchronous SCSI command
is only active if USE_IMMED is 1. It wrongly expects wait_cmd() to fill
its submitted sense data buffer and it determines the end of the
asynchronous command from the unreliable progress messages of the drive.

So replace functions print_completion_info() and sig_progress() by a new
function wait_for_unit_ready() and enable USE_IMMED.

Signed-off-by: Thomas Schmitt <scdbackup@gmx.net>

scdbackup committed:
cdrwtool: Switch to FORMAT UNIT Format Code 1

The specs from MMC-1 to MMC-3 prescribe for FORMAT UNIT on CD-RW media the
Format Code 7 and the related format descriptor layout. Beginning with
MMC-4 this code and layout have been moved to Legacy Specifications.
Pioneer BD-RW BDR-S09 rev 1.51 refuses on this code but works with the
more modern Format Code 1 with Format Type 10h "CD-RW/DVD-RW Full Format".

So introduce a new function format_disc_code1() and rename the old
function format_disc() to format_disc_code7(). Create a new format_disc()
which first tries format_disc_code1() and in case of failure tries
format_disc_code7().

Signed-off-by: Thomas Schmitt <scdbackup@gmx.net>

scdbackup commented:
Hi,

renaming "v2" to "master" does not work because 

    $ hub push scdbackup
    ...
     ! [rejected]        master -> master (non-fast-forward)
    error: failed to push some refs to 'https://github.com/scdbackup/udftools.git'
    hint: Updates were rejected because the tip of your current branch is behind
    hint: its remote counterpart. Integrate the remote changes (e.g.
    hint: 'git pull ...') before pushing again.
    hint: See the 'Note about fast-forwards' in 'git push --help' for details.

I tried pushing after

    git branch -m master old_master

but that seems to only have pushed "v2" and tells me

    remote: Create a pull request for 'v2' on GitHub by visiting:
    remote:      https://github.com/scdbackup/udftools/pull/new/v2

After

    git branch -m v2 master

i get above rejection at a push attempt.
So i rename everything back and push to get told "Everything up-to-date".

At "checkout branch master" i get told

    Your branch is ahead of 'origin/master' by 5 commits.
    (use "git push" to publish your local commits)

So it looks like i managed to get near to capsizing.
But i see the commits in https://github.com/scdbackup/udftools/commits .
I try to push (what do i have to lose with patches in my backups ?)
but it says "Everything up-to-date".

I tried "git pull" and get "Already up to date.".

I applied and committed my improved patches, made sure the results match the
canned files from "master", and tried to push:

    ! [rejected]        master -> master (non-fast-forward)
    error: failed to push some refs to 'https://github.com/scdbackup/udftools.git'

I am stuck. This attempt of renaming seems to have been deadly.

Shall i try to issue a pull request for branch "v2" ?
It seems to exist:
  https://github.com/scdbackup/udftools/commits/v2

Have a nice day :)

Thomas

pali commented:
> renaming "v2" to "master" does not work because 
> 
>     $ hub push scdbackup
>     ...
>      ! [rejected]        master -> master (non-fast-forward)
>     error: failed to push some refs to 'https://github.com/scdbackup/udftools.git'

Because you have not used `--force` argument.

> Shall i try to issue a pull request for branch "v2" ?

No, the "rule" is to use one pull requesy for one change. And not "many" pull requests for one change (or many unrelated changes in one pull request).

Just to note: Github show there what was changed between v1 and v2 and so it helps people to do review as they quickly see what is important what not. Opening a new pull request means that "reviewer" needs to start from beginning again and all details, description, communication is split into two github pages, which makes thing hard...

scdbackup commented:
Hi,

    $ hub push --force scdbackup
    Username for 'https://github.com': scdbackup@gmx.net
    Password for 'https://scdbackup@gmx.net@github.com':
    Everything up-to-date

    $ git push --force scdbackup
    Username for 'https://github.com': scdbackup@gmx.net
    Password for 'https://scdbackup@gmx.net@github.com':
    Everything up-to-date

Still 5 commits to see in
  https://github.com/scdbackup/udftools/commits/master
whereas my local master's git log shows only two of them.

("Schatz ! Ich hab' das Internet geloescht !"
= "Darling ! I just deleted the Internet !")

Have a nice day :)

Thomas

pali commented:
> Everything up-to-date
> Still 5 commits to see in
> https://github.com/scdbackup/udftools/commits/master
> whereas my local master's git log shows only two of them.

And on which local branch you are currently? `git branch` will show it. And what to which remote branch will `git push` (without arguments) upload changes from your current local branch? File `.git/config` will tell you it (see section branch).

I guess that default was trying to push your local changes to v2 branch; not to master...

You can use full-syntax of specifying local and remote branch for `git push` as I wrote in comment https://github.com/pali/udftools/issues/33#issuecomment-751682050 So you can esnure that you push changes into master branch on github.

scdbackup commented:
Hi,

> And on which local branch you are currently?

"master", stemming from "git reset", "git apply", and "git commit"
of the two revised patches. I now gave it a round-trip over
"git checkout v2" and "git checkout old_master", a branch which i forked
off the old "master" before resetting it. Then back to
"git checkout master".

Oh miracle:

    $ hub push --force scdbackup
    Username for 'https://github.com': scdbackup@gmx.net
    Password for 'https://scdbackup@gmx.net@github.com': 
    Enumerating objects: 16, done.
    Counting objects: 100% (16/16), done.
    Delta compression using up to 8 threads
    Compressing objects: 100% (11/11), done.
    Writing objects: 100% (11/11), 3.22 KiB | 3.22 MiB/s, done.
    Total 11 (delta 7), reused 0 (delta 0)
    remote: Resolving deltas: 100% (7/7), completed with 4 local objects.
    To https://github.com/scdbackup/udftools.git
     + d145cdd...8854a98 master -> master (forced update)

Now the remote "master" has two commits.
(Given its history i would be extra mistrusting towards its content.)

> You can use full-syntax of specifying local and remote branch for git push

I doubt that a wrong target was the final problem with
"git push scdbackup". It appeared that remote perception of the "master"
branch partly knew about the local changes and partly knew not.
Rattling the tree seems to have helped.
But i may be wrong ...

Have a nice day :)

Thomas

pali commented:
> scdbackup force-pushed the scdbackup:master  branch from d145cdd to 8854a98 19 minutes ago

It is updated, github wrote above note.

Great!

pali commented:
Thanks! Now I merged it.

I'm planning to do tests and then release a new version of udftools. If you have something which could like to fix/change let me know.

scdbackup commented:
Hi,

thanks for your patience with my difficulties to interact with GitHub.

> If you have something which could like to fix/change let me know.

Since Travis is part of the development process, i think that the
Makefiles should use -Wall -Werror as compiler arguments in order to
show a similar pickiness as Travis. I'm all in for -Wall, but did not
pay attention to the fact that it was not active during development of
my changes.

The message at the end of a cdrwtool run "can't unlock door" is ugly
but does not hamper functionality, as far as i can see. The drive tray
can be manually ejected afterwards. It does not look like cdrwtool would
try to eject from inside.
Replacing the tray related CDROM ioctls by SCSI commands would be 
worthwhile, but imposes the traditional risks of repairing things which 
are not really broken.

Have a nice day :)

Thomas

pali commented:
-Werror should not be used as we do not know which new warnings appear in new compiler versions in future... and this may cause fully-warning-free code failed to compile in future.

udftools is using autotools and idea of autotools is that caller sets compile flags as wish...

scdbackup commented:
Hi,

> -Werror should not be used as we do not know which new warnings appear in
> new compiler versions in future

I think that at least -Wall should be used by default.

--------------------------------------------------------------------------

I am pondering where to mention the (still riddling) fact that 4 of 5
tested drives (1 of 2 of callegar and all my 3) throw error on
set_write_mode()'s MODE SELECT command with CD-RW unless the medium is
written and closed:

    Command failed: 55 10 00 00 00 00 00 00 3c 00 00 00 - sense 05.26.00

Would a "BUGS" section in doc/cdrwtool.1 be appropriate ?
It's not yet to be considered a bug of the program, but rather a problem
dangling between program and drive firmwares.

    BUGS
         Many modern drives refuse on the preparations to format new,
         blanked, or appendable CD-RW media. This causes a message like
           "Command failed: 55 ... - sense ..."
         The remedy is to use a CD-capable burn program to write a session
         and closing the medium. For example by using any of "cdrecord",
         "wodim", "cdrskin", or "xorriso -as cdrecord" as content of
         variable "prog" in:
           prog="xorriso -as cdrecord"
           drive="/dev/sr0"
           dd if=/dev/zero bs=1M count=10 | $prog -v -eject dev="$drive" -

The definition of "BUGS" in man 7 man-pages would include this diffuse
kind of problem:

    A list of limitations, known defects or inconveniences,
    and other questionable activities.

Have a nice day :)

Thomas

pali commented:
BUGS section in cdrwtool manpage is ok. Are you going to create a pull request with it? :-)

scdbackup commented:
Hi,

> Are you going to create a pull request with it? :-)

Do i have a choice ? :))

I will try to learn how to get scdbackup/udftools in sync with pali/udftools
and then replay the good part of what i logged in issue 33 and here.

What i hope to achieve is:
One branch - one commit - one pull request - maybe more commits - 
finally a simple squashing of everything that's new in the branch.

Have a nice day :)

Thomas

pali commented:
> I will try to learn how to get scdbackup/udftools in sync with pali/udftools
> and then replay the good part of what i logged in issue 33 and here.

Check that in .git/config you have defined two remotes. One for pali/udftools (sometimes called `upstream` or by name `pali`) and one for scdbackup/udftools (sometimes called `origin` or by name `scdbackup`).

Then create two local branches, one which will be linked with pali/udftools and one with scdbackup/udftools. Do it via `git branch --set-upstream-to=` for currently checkout branch.

After that calling `git pull` and `git push` on checkouted branch would do synchronization with specified tracked remote branch on github.

To "sync" pali/udftools into scdbackup/udftools, just checkout your local branch for scdbackup/udftools, call `git merge --ff-only <branch>` with branch which points to the pali/udftools.

pali commented:
Now I tested cdrwtool with all these patches... formatting did not show me any error, but formated CD-RW disc is not suitable for kernel pktcdvd.ko driver. Problem is that READ_DISC_INFO reports that _erasable_ bit is not set. And kernel pktcdvd.ko driver requires it for CD-RW disc.
```
$ ./cdrwtool/cdrwtool -d /dev/sr0 -t 1 -q
using device /dev/sr0
setting speed to 1
256KB internal buffer
setting write speed to 1x
Settings for /dev/sr0:
        Fixed packets, size 32
        Mode-2 disc

I'm going to do a quick setup of /dev/sr0. The disc is going to be blanked and formatted with one big track. All data on the device will be lost!! Press CTRL-C to cancel now.
ENTER to continue.

Initiating quick disc blank
00% complete
07% complete
39% complete
71% complete
100% complete and done
Disc capacity is 295264 blocks (590528KB/576MB)
Formatting track
00% complete
01% complete
02% complete
03% complete
04% complete
05% complete
06% complete
07% complete
08% complete
09% complete
10% complete
11% complete
12% complete
13% complete
14% complete
15% complete
16% complete
17% complete
18% complete
19% complete
20% complete
21% complete
22% complete
23% complete
25% complete
25% complete
26% complete
27% complete
28% complete
29% complete
30% complete
32% complete
32% complete
33% complete
35% complete
35% complete
37% complete
38% complete
39% complete
39% complete
41% complete
42% complete
42% complete
44% complete
45% complete
46% complete
47% complete
48% complete
49% complete
50% complete
51% complete
52% complete
53% complete
54% complete
55% complete
56% complete
57% complete
58% complete
59% complete
60% complete
61% complete
62% complete
63% complete
64% complete
65% complete
66% complete
67% complete
68% complete
69% complete
70% complete
71% complete
72% complete
73% complete
74% complete
75% complete
76% complete
77% complete
78% complete
79% complete
80% complete
82% complete
82% complete
83% complete
84% complete
85% complete
86% complete
87% complete
89% complete
89% complete
91% complete
91% complete
92% complete
94% complete
94% complete
94% complete
95% complete
96% complete
96% complete
98% complete
98% complete
100% complete and done
start=0, blocks=16, type=RESERVED 
start=16, blocks=4, type=VRS 
start=20, blocks=76, type=USPACE 
start=96, blocks=32, type=MVDS 
start=128, blocks=32, type=LVID 
start=160, blocks=5, type=STABLE 
start=165, blocks=91, type=USPACE 
start=256, blocks=1, type=ANCHOR 
start=257, blocks=31, type=USPACE 
start=288, blocks=1024, type=SSPACE 
start=1312, blocks=293664, type=PSPACE 
start=294976, blocks=31, type=USPACE 
start=295007, blocks=1, type=ANCHOR 
start=295008, blocks=96, type=USPACE 
start=295104, blocks=32, type=RVDS 
start=295136, blocks=96, type=USPACE 
start=295232, blocks=5, type=STABLE 
start=295237, blocks=26, type=USPACE 
start=295263, blocks=1, type=ANCHOR 
Writing UDF structures to disc
Quick setup complete!
$ sync
$ eject /dev/sr0
```
... insert again ...
```
$ ./pktsetup/pktcdvd-check /dev/sr0 
CD-RW device '/dev/sr0' is not formatted to erasable mode
```

```
$ ./pktsetup/pktsetup /dev/sr0
kernel: [73914.198319] pktcdvd: pktcdvd0: writer mapped to sr0
```
```
$ ./mkudffs/mkudffs /dev/pktcdvd/pktcdvd0
Detected CD-RW optical disc
mkudffs: Error: Cannot open device '/dev/pktcdvd/pktcdvd0': Read-only file system
kernel: [73926.451430] pktcdvd: pktcdvd0: can't write to this track
```

scdbackup commented:
Hi,

> $ ./pktsetup/pktcdvd-check /dev/sr0 
> CD-RW device '/dev/sr0' is not formatted to erasable mode

Ah. Hacking. Away with the git docs !

Try this:

    diff --git a/pktsetup/pktcdvd-check.c b/pktsetup/pktcdvd-check.c
    index 28705a5..e293a4f 100644
    --- a/pktsetup/pktcdvd-check.c
    +++ b/pktsetup/pktcdvd-check.c
    @@ -188,7 +188,7 @@ int main(int argc, char *argv[]) {
     
            memset(&cgc, 0, sizeof(cgc));
            memset(&sense, 0, sizeof(sense));
    -       memset(&discinfo, 0, sizeof(discinfo));
    +       memset(&trackinfo, 0, sizeof(trackinfo));
            cgc.cmd[0] = GPCMD_READ_TRACK_RZONE_INFO;
            cgc.cmd[1] = 1; /* type: 1 */
            cgc.cmd[5] = 1; /* track: 1 */

For me it causes

    $ ./pktsetup/pktcdvd-check /dev/sr0
    Optical device '/dev/sr0' is compatible with pktcdvd

-------------------------------------------------------------------------

Long story:

On a CD-RW which was formatted during my tests i see with

    $ xorriso -scsi_log on -outdev /dev/sr0 2>&1 | less

this transaction

    READ DISC INFORMATION
    51 00 00 00 00 00 00 00 22 00 
    From drive: 34b
    00 20 1e 01 01 01 01 80 20 00 00 00 00 69 14 84 ff ff ff ff
    ff ff ff ff 00 00 00 00 00 00 00 00 00 00 

The Erasable bit is number 4 in byte 2 = 0x1e. So it is on.

I am not aware that it can be set to 0 by any SCSI operation.
MMC-5 says:

    6.22.3.1.2 Erasable Bit
    The Erasable bit, when set to one, indicates that CD-RW, DVD-RAM,
    DVD-RW, DVD+RW, HD DVD-RAM, or BD-RE media is present and the Drive
    is capable of writing the media. If the Erasable bit is set to zero,
    then either the medium is not erasable or the Drive is unable to
    write the media.

Now for pktcdvd-check. Mine complains like yours.
That message does not reflect the situation, though, because the bit has
nothing to do with formatting but rather with the medium profile, i.e.
the capabilities of the drive and the medium type.

I just wanted to insert printfs() when i see this in line 194, after
discinfo was filled with the full info in line 184:

            memset(&discinfo, 0, sizeof(discinfo));
            cgc.cmd[0] = GPCMD_READ_TRACK_RZONE_INFO;

So line 239 inquires a zeroised discinfo, rather than the reply from
the drive.

(Will next look into kernel writability ...)

Have a nice day :)

Thomas

pali commented:
> Will next look into kernel writability ...

Kernel is using same checks as pktcdvd-check. Or rather I have tried to write pktcdvd-check to call same commands as kernel.

scdbackup commented:
Hi,

> $ ./mkudffs/mkudffs /dev/pktcdvd/pktcdvd0
> Detected CD-RW optical disc
> mkudffs: Error: Cannot open device '/dev/pktcdvd/pktcdvd0': Read-only file s
ystem

Try:

    diff --git a/mkudffs/main.c b/mkudffs/main.c
    index a384ea8..c3e22ce 100644
    --- a/mkudffs/main.c
    +++ b/mkudffs/main.c
    @@ -408,6 +408,8 @@ int main(int argc, char *argv[])
                    if (S_ISBLK(stat.st_mode))
                            flags2 |= O_EXCL;
     
    +               close(fd);
    +               fd = -1;
                    fd2 = open(filename2, flags2);
                    if (fd2 < 0)
                    {
    @@ -421,8 +423,6 @@ int main(int argc, char *argv[])
                                    exit(1);
                            }
                    }
    -
    -               close(fd);
                    fd = fd2;
            }

---------------------------------------------------------------------------

The medium appears to be random access writable.
After making a backup of block 12345 of /dev/sr0 i can do:

    $ dd if=/dev/zero bs=2048 count=1 seek=12345 of=/dev/sr0
    1+0 records in
    1+0 records out
    2048 bytes (2.0 kB, 2.0 KiB) copied, 0.000301614 s, 6.8 MB/s
    $ echo "Hello block 12345" | dd bs=2048 count=1 seek=12345 of=/dev/sr0
    0+1 records in
    0+1 records out
    18 bytes copied, 0.000152535 s, 118 kB/s

The data come back on request:

    $ (dd if=/dev/sr0 bs=2048 count=1 skip=12345 | dd bs=1 count=18 ) 2>/dev/null
    Hello block 12345

I restore the backuped block 12345 (which was entirely zero) and do:

    $ ./mkudffs/mkudffs /dev/pktcdvd/pktcdvd0
    Detected CD-RW optical disc
    mkudffs: Error: Cannot open device '/dev/pktcdvd/pktcdvd0': Device is mounted or mkudffs is already running
    $ sudo umount /dev/pktcdvd/pktcdvd0
    umount: /dev/pktcdvd/pktcdvd0: not mounted.
    $ sudo umount /dev/sr0
    umount: /dev/sr0: not mounted.

---------------------------------------------------------------------------

Something seems to be wrong with the O_EXCL protocol of
/dev/pktcdvd/pktcdvd0 . I can reproduce the EBUSY failure of open(2) by

    fd= open(name, O_RDONLY);
    fd2= open(name, O_RDWR | O_EXCL);
    if (fd2 == -1)
            perror("open(O_RDWR | O_EXCL)");
    else
            printf("success\n");

With name="/dev/sr0" this yields "success".
With name="/dev/pktcdvd/pktcdvd0" i get:

    open(O_RDWR | O_EXCL): Device or resource busy

If i leave out the first open(), then i get "success".

So mkudffs/main.c has to close fd before opening fd2.
After changing mkudffs/main.c i get 

    $ ./mkudffs/mkudffs /dev/pktcdvd/pktcdvd0
    Detected CD-RW optical disc
    filename=/dev/pktcdvd/pktcdvd0
    label=LinuxUDF
    uuid=5ff0de4b977a3aea
    blocksize=2048
    blocks=295264
    udfrev=2.01
    start=0, blocks=16, type=RESERVED 
    ...
    start=295237, blocks=26, type=USPACE 
    start=295263, blocks=1, type=ANCHOR 
    $

Have a nice day :)

Thomas

scdbackup commented:
Hi,

i see that the mkudffs diff fixes my situation but not yours.

What do you get from dd experiments ?
If failure: Do you have w-permission on both involved devices ?

Have a nice day :)

Thomas

pali commented:
>             memset(&discinfo, 0, sizeof(discinfo));

This is really stupid error :-( Thank you for finding it. Should be fixed in commit 2ab29480d4c7d5dabb282e3895b60eb3cbac280b.

pali commented:
I already removed whole O_EXCL hacks as it does not work.

pali commented:
Now pktcdvd-check show that trackinfo's fixed packet (fp) bit is not set. And kernel complains too:
```
$ ./mkudffs/mkudffs /dev/pktcdvd/pktcdvd0 
Detected CD-RW optical disc
mkudffs: Error: Cannot open device '/dev/pktcdvd/pktcdvd0': Read-only file system
```
```
kernel: [87185.154496] pktcdvd: pktcdvd0: can't write to this track
```
strace shows:
```
openat(AT_FDCWD, "/dev/pktcdvd/pktcdvd0", O_RDWR|O_EXCL) = -1 EROFS (Read-only file system)
```

scdbackup commented:
Hi,

i get "Read-only file system" only with an unformatted CD-RW and only
if i open it with O_RDWR. (O_EXCL or not does not matter.)

What do you get from this inspection of the medium:

    xorriso -outdev /dev/sr0 -toc

A formatted one gets reported with

    Media status : is written , is closed
    Media blocks : 295264 readable , 0 writable , 359847 overall

An unformatted one will probably show a different "readable" number.

Do you know where the kernel distinguishes both states ?

Have a nice day :)

Thomas

pali commented:
This is interesting... it does nto contain TOC even cdrwtool formatted it and written UDF fs on it...
```
Media status : is written , is closed
Media blocks : 1 readable , 0 writable , 359847 overall
TOC layout   : Idx ,  sbsector ,       Size , Volume Id
xorriso : SORRY : Cannot obtain Table Of Content
xorriso : NOTE : Tolerated problem event of severity 'SORRY'
xorriso : NOTE : -return_with SORRY 32 triggered by problem severity SORRY
```

pali commented:
> Do you know where the kernel distinguishes both states ?

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/block/pktcdvd.c?h=v5.10#n1708
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/block/pktcdvd.c?h=v5.10#n1675

scdbackup commented:
Hi,

> Now pktcdvd-check show that trackinfo's fixed packet (fp) bit is not set.

That might be the indicator which the kernel uses for random writability.

I get from "xorriso -scsi_log on -outdev /dev/sr0" with the formatted
CD-RW:

    READ TRACK INFORMATION
    52 01 00 00 00 01 00 00 22 00 
    From drive: 34b
    00 22 01 01 00 07 b2 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 20 00 04 81 60 00 00 00 00 00 00 

FP is bit4 in byte 6 = 0xb2, i.e. set to 1.

With the unformatted and closed CD-RW, the reply is:

    00 22 01 01 00 04 01 00 00 00 00 00 00 00 00 00 00 00 00 00
    00 00 00 00 00 00 85 81 00 00 00 00 00 00 

FP is not set.
Another interesting difference is in byte 5, which says that the formatted
CD-RW has a "Data track, recorded increment", whereas the unformatted CD-RW
has a "Data track, recorded uninterupted".

> https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/block/pktcdvd.c?h=v5.10#n1675

Ahum.

    if (!ti->packet || !ti->fp)
            return 0;

That's byte 6 bit 5 and byte 6 bit 7.
Both are set in READ TRACK INFORMATION from the formatted CD-RW and not set
in the reply from the unformatted CD-RW.

> xorriso : SORRY : Cannot obtain Table Of Content

Can you send the full output of

    xorriso -scsi_log on -outdev /dev/sr0 -toc 2>&1 | tee -i /tmp/xorriso.log

directly to scdbackup@gmx.net ?

Have a nice day :)

Thomas

scdbackup commented:
Hi,

the SCSI log of the bad CD-RW has arrived by mail.
The decisive failed commands for getting a Table-Of-Content are:

    READ TOC/PMA/ATIP
    43 02 02 00 00 00 00 00 04 00
    +++ sense data = 70 00 05 00 00 00 00 0A 00 00 00 00 24 00 00 00 00 00
    +++ key=5  asc=24h  ascq=00h
       27976 us     [ 1572164 ]

This tries to inquire a raw CD TOC. The error reply is

    5 24 00 INVALID FIELD IN CDB

With my formatted CD-RW i get the available reply size

    READ TOC/PMA/ATIP
    43 02 02 00 00 00 00 00 04 00
    From drive: 4b
    00 4f 01 01
        7427 us     [ 2408625 ]

and can then request these data

    READ TOC/PMA/ATIP
    43 02 02 00 00 00 00 00 51 00
    From drive: 81b
    00 4f 01 01 01 17 00 a0 00 00 00 00 01 20 00 01 16 00 a1 00
    00 00 00 01 00 00 01 17 00 a2 00 00 00 00 4f 3b 49 01 17 00
    01 00 00 00 00 00 02 00 01 57 00 b0 ff ff ff 03 4f 3b 4a 01
    57 00 c0 66 00 9c 00 61 1a 41 01 57 00 c1 48 34 78 00 00 00
    00
         6538 us     [ 2415453 ]

Then libburn tries to whether the medium is readable by requesting
one block beginning at LBA 0:

    READ(10)
    28 00 00 00 00 00 00 00 01 00
    +++ sense data = 70 00 05 00 00 00 00 0A 00 A9 00 00 21 00 00 00 00 00
    +++ key=5  asc=21h  ascq=00h
        31471 us     [ 1604127 ]

This error reply means

    5 21 00 LOGICAL BLOCK ADDRESS OUT OF RANGE

I.e. the drive thinks that block 0 is not readable.

Now i am puzzled in what ill state your medium might be.
Do you have another CD-RW medium for trying ?

Have a nice day :)

Thomas

pali commented:
Immediately after formatting this disc (without ejecting it) is pktcdvd-check returning that it is compatible and also kernel can open it for R/W mode. But trying to write to it just ends in kernel errors:
```
kernel: [94400.959524] sr 1:0:0:0: [sr0] tag#25 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
kernel: [94400.959539] sr 1:0:0:0: [sr0] tag#25 Sense Key : Hardware Error [current]
kernel: [94400.959545] sr 1:0:0:0: [sr0] tag#25 Add. Sense: Track following error
kernel: [94400.959552] sr 1:0:0:0: [sr0] tag#25 CDB: Read(10) 28 00 00 00 00 00 00 00 10 00
kernel: [94400.959557] print_req_error: I/O error, dev sr0, sector 0
```
So seems that this CD-RW medium is broken/damaged (maybe too many erase cycles)... At least it helped to discover these issues in mkudffs and pktcdvd-check :-)

scdbackup commented:
Hi,

> Immediately after formatting this disc (without ejecting it) is
> pktcdvd-check returning that it is compatible and also kernel can open it
> for R/W mode.

This looks like the Table-Of-Content was not properly written to the
medium. The drive was willing to work with it as long as it memorized
what it supposedly wrote there. But the assessment after loading the
medium yields no usable state.

If there is no other CD-RW at hand, consider to blank it thoroughly and
to write it fully before trying cdrwtool again:

    dd if=/dev/zero | xorriso -as cdrecord -v dev=/dev/sr0 blank=all -

The burn run  is supposed to last about two full writing times and to end
by messages like:

    ...
    libburn : FATAL : SCSI error on write(359840,16): [5 21 02] Illegal request. Invalid address for write.
    ...
    xorriso : UPDATE : Closing track/session. Working since 488 seconds
    ...
    libburn : FATAL : Burn run failed
    xorriso : FAILURE : libburn indicates failure with writing.
    xorriso : NOTE : Gave up -outdev ''
    xorriso : FAILURE : -as cdrecord: Job could not be performed properly.
    xorriso : aborting : -abort_on 'FAILURE' encountered 'FATAL'

This error message is a consequence of dd shoveling an unlimited amount
of data. This run 

    dd if=/dev/zero bs=2048 count=359847 | \
    xorriso -as cdrecord -v dev=/dev/sr0 blank=all padsize=0 -

would curb it to what the drive partly told as its capacity. But currently
i don't believe what the drive tells about this medium.

When the full blank and burn is done, try to format the medium and check
whether the formatted state survives eject and reload.

Have a nice day :)

Thomas

scdbackup commented:
Hi,

it turned out that my local situation is not as much tangled as i thought.
"master" is associated to pali/udftools "master" which i could pull after
revoking my uncommitted hacks of yesterday.
I branched "man_cdrwtool_needs_closed", committed to it, and pushed to
scdbackup/udftools by the locally configured target repository "scdbackup".
The commit is visible in
    https://github.com/scdbackup/udftools/commits/man_cdrwtool_needs_closed

But the next step fails

    $ hub pull-request -m 'doc: Mention in man cdrwtool the workaround for MODE SELECT failure on CD-RW'
    github.com username: ...
    github.com password for scdbackup@gmx.net (never stored):
    Error creating pull request: Not Found (HTTP 404)
    Not Found

Is this because there is no branch "man_cdrwtool_needs_closed" in
pali/udftools ?

(I try to wrap my head around your prescription
> Then create two local branches, one which will be linked with pali/udftools
> and one with scdbackup/udftools. Do it via git branch --set-upstream-to= for
> currently checkout branch.
> After that calling git pull and git push on checkouted branch would do
> synchronization with specified tracked remote branch on github.
but pushing the new branch to "origin" or
"https://github.com/pali/udftools.git" fails.)

Have a nice day :)

Thomas

pali commented:
`hub pull-request` has options: `--base=BASE --head=HEAD`. Try to use `--base=pali:master --head=scdbackup:man_cdrwtool_needs_closed`

scdbackup commented:
Hi,

no success:

    $ hub pull-request --base=pali:master --head=scdbackup:man_cdrwtool_needs_closed -m 'doc: Mention in man cdrwtool the workaround for MODE SELECT failure on CD-RW'
    github.com username: ...
    github.com password for scdbackup@gmx.net (never stored):
    Error creating pull request: Not Found (HTTP 404)
    Not Found

Have a nice day :)

Thomas

pali commented:
Are you using github token or password? (Because for hub operations you need to use token and github returns 404 for any error, including "incorrect token"). You can also configure token in `~/.config/hub` file so it would not be asked every time.

scdbackup commented:
Hi,

no success with:

    $ hub pull-request --base=pali:master --head=scdbackup:man_cdrwtool_needs_closed -m 'doc: Mention in man cdrwtool the workaround for MODE SELECT failure on CD-RW'
    github.com username: ...
    github.com password for scdbackup@gmx.net (never stored):
    Error creating pull request: Not Found (HTTP 404)
    Not Found

In my despair i went to the URL, which "hub push scdbackup" advised:

    https://github.com/scdbackup/udftools/pull/new/man_cdrwtool_needs_closed

and clicked the green button "Create pull request".
Now there is https://github.com/pali/udftools/pull/45

Have a nice day :)

Thomas

scdbackup commented:
Hi,

> Are you using github token or password?

Hrmpf. Yes. I forgot to set the GITHUB_TOKEN= environment variable.
"hub" anyways asks me for the password. So i saw no difference in
behavior. The variable was set and exported when i had my successful push 
to "master".

Whatever, the push seems to have succeeded via web interface (where i
don't get asked any password or token). So for now i cannot make more 
experiments.

Have a nice day :)

Thomas

pali commented:
```
$ dd if=/dev/zero | xorriso -as cdrecord -v dev=/dev/sr0 blank=all -
xorriso 1.5.0 : RockRidge filesystem manipulator, libburnia project.

xorriso : NOTE : Disc status unsuitable for writing
Drive current: -outdev '/dev/sr0'
Media current: CD-RW
Media status : is written , is closed
Beginning to blank medium in mode 'all'.
libburn : SORRY : Drive and media state unsuitable for blanking. (role= 1 , profile= 0xa , status= 6)
xorriso : FAILURE : Blanking failed.
xorriso : NOTE : Re-assessing -outdev '/dev/sr0'
xorriso : NOTE : Disc status unsuitable for writing
Drive current: -outdev '/dev/sr0'
Media current: CD-RW
Media status : is written , is closed
xorriso : FAILURE : Cannot recognize state of drive and media
xorriso : NOTE : Gave up -outdev ''
xorriso : FAILURE : -as cdrecord: Job could not be performed properly.
xorriso : aborting : -abort_on 'FAILURE' encountered 'FAILURE'
```
Seems that this CD-RW disc is totally dead.

> Do you have another CD-RW medium for trying ?

I cannot find any other, only CD-R. So I would have to buy some new.

Anyway, cdrwtool still can format this disc again and again without failure. I tested also `format_disc_code7` code path and it is working fine. And also `format_disc_code7` with `USE_IMMED=0` is working fine (no progressbar, just waiting) and also `format_disc_code1` with `USE_IMMED=0` `USE_IMMED_WITH_FORMAT_1=0`. So my dvd burner accepts any combination of mmc commands.

scdbackup commented:
Hi,

> xorriso : NOTE : Disc status unsuitable for writing

Probably because the Table-Of-Content cannot be read.
Changing "-v" to "-vvv" would probably yield a program message about
"Could not inquire TOC".

> Seems that this CD-RW disc is totally dead.

You could try with wodim or cdrecord instead of xorriso (cdrskin uses
libburn like xorriso does). Maybe they have less scruples with bad TOC.

Have a nice day :)

Thomas
